[sil-optimizer] Add FP comparison support in constant folder

jkshtj opened the pull request:
This pull request adds support for handling floating point comparisons to the constant folder. 

Partially fixes #68901

jkshtj committed:
[sil-optimizer] Add FP comparison support in constant folder

jkshtj commented:
@BradLarson could you please trigger swift-ci?

jkshtj commented:
@eeckstein could you please take a look at the PR?

asl commented on lib/SILOptimizer/Utils/ConstantFolding.cpp:
```diff
@@ -50,7 +50,66 @@ APInt swift::constantFoldBitOperation(APInt lhs, APInt rhs, BuiltinValueKind ID)
+  bool result;
+  bool isOrdered = !lhs.isNaN() && !rhs.isNaN();
+
+  switch (ID) {
```
Please ensure your coding style is same as in the file

jkshtj replied:
This is actually `git-clang-format`. But I can bypass that step to keep the style same.

asl replied:
Well... `git-clang-format` is dumb and overall the style guide is not _simply clang-format everything_.

jkshtj replied:
> not simply clang-format everything.

Shouldn't it be? What is the benefit of using a formatter then?

asl replied:
It can handle maybe 80% of cases. However, the remaining 20% of cases are about things like readability, etc.

However, in your particular case it's not configured properly. The repo's style config is explicitly said `based on LLVM`. And LLVM defaults to 2 space indent. So, something is not correct here and this is why one need to examine the output of formatter not merely commit it as-is.

asl commented on test/AutoDiff/SILOptimizer/vjp_and_pullback_inlining.swift:
```diff
@@ -40,7 +40,12 @@ func caller_of_simple_vjp() -> Float {
 @_silgen_name("pb_with_control_flow")
 func pb_with_control_flow(_ x: Float) -> Float {
   if (x > 0) {
-    return sin(x) * cos(x)
```
I would just ensure we won't have constant `x` here. For example, you could have an argument of `caller_of_pb_with_control_flow` that you will use to evaluate gradient at.

jkshtj replied:
Ah nice catch. Fixed.

eeckstein reviewed (commented):
Nice!
Do you mind adding a swift (not sil) test file which compares constant folded results with not-constant folded floating point operations? This would not only test the implementation but also the assumptions about the floating point behavior (like "Ordered comparisons with NaN always return false", etc).

You can do that for example by generating the test file, like e.g. done here: https://github.com/apple/swift/blob/main/validation-test/SILOptimizer/static_enums_fuzzing.swift.
Then it's easy to test all combinations of fp-types * operations * relevant-operand-values

jkshtj committed:
[sil-optimizer] Add validation tests for constant folded FP arithmetic

jkshtj commented:
@eeckstein Added the validation tests as you suggested. They helped catch quite a lot of edge cases so thanks for pointing it out here!

Can you please take a look at the revision when you get time?

eeckstein commented on validation-test/SILOptimizer/constant_folded_fp_operations.swift:
```diff
@@ -0,0 +1,633 @@
+        }
+    }
+
+        func generateComparisonFuncDecls() {
```
wrong indentation

jkshtj replied:
Fixed.

eeckstein commented on validation-test/SILOptimizer/constant_folded_fp_operations.swift:
```diff
@@ -0,0 +1,633 @@
+                        @inline(never) @_silgen_name("\(comparisonFuncName)") @_optimize(none)
+                        func \(comparisonFuncName)() {
+                            let \(resultVarName) = \(optFuncName)() == \(unoptFuncName)()
+                            print("[\(resultVarName)] result equal? \\(\(resultVarName))")
```
You could just emit a `precondition(result, "message")` instead of FileChecking the output. But checking the output is also fine

jkshtj replied:
Nice! Changed.

eeckstein reviewed (approved):
Very nice! Thanks!
LGTM

jkshtj commented:
@eeckstein we may need to restart swift-ci.

asl commented:
@swift-ci please test

jkshtj commented:
Looks like a validation test failed. I'm taking a look.

jkshtj committed:
[sil-optimizer] Don't constant fold eq comparisons b/w inf and non-inf

This commit modifes the constant folder to not handle equality comparisons b/w
infinity and non-infinity operands. In such comparisons, special floating point
types - Float80 and Float16, may come into play and pattern matching againt them
complicates the constant folding logic more than we'd like.

jkshtj commented:
Fixed the failing validation test.

> This commit modifes the constant folder to not handle equality comparisons b/w
    infinity and non-infinity operands. In such comparisons, special floating point
    types - Float80 and Float16, may come into play and pattern matching againt them
    complicates the constant folding logic more than we'd like.

BradLarson commented:
@swift-ci Please test

jkshtj commented:
@eeckstein I believe this can be merged?

eeckstein commented:
Basically yes.
Please rebase + retest it first

jkshtj commented:
@eeckstein Here's my changes rebased on top of trunk. Can we kick off swift-ci?

eeckstein commented:
@swift-ci test

jkshtj commented:
Windows build seems to have timed out.

asl commented:
@swift-ci please test windows

tbkka commented:
@jkshtj -- Do the LLVM optimizations not handle these cases?  Generally, LLVM is pretty good at this sort of thing.

jkshtj commented:
> @jkshtj -- Do the LLVM optimizations not handle these cases? Generally, LLVM is pretty good at this sort of thing.

@tbkka we were motivated to perform FP constant-folding at the SIL level because it vastly simplified the generated derivative code. Below is a minimal example of the optimizations we wanted the FP constant-folding to enable.
```swift 
@differentiable(reverse)
func f(_ x: Float) -> Float {
  if (x > 0) {
    return sin(x) * cos(x)
  } else {
    return sin(x) + cos(x)
  }
}

// With FP constant-folding the control-flow in `f` 
// is removed after it's inlined into `foo`, which eventually 
// enables further inlining of all VJPs and pullbacks involved.
//
// Inlining of VJPs and pullbacks lets us get rid of closure context
// allocations, which form the most latent chunk of AD code.
@inline(never)
func foo() -> Float {
  gradient(at: Float(4), of: f(x) )
}
```
